net/udprelay/server.go

```diff
@@ -0,0 +1,517 @@
+	// Disco protocol messages for transmission towards discoPubKeys[0].
+	discoPubKeys       pairOfDiscoPubKeys
+	discoSharedSecrets [2]key.DiscoShared
+	handshakeState     [2]disco.BindUDPRelayHandshakeState
+	addrPorts          [2]netip.AddrPort
+	lastSeen           [2]time.Time
+	challenge          [2][disco.BindUDPRelayEndpointChallengeLen]byte
+
+	lamportID   uint64
+	vni         uint32
+	allocatedAt time.Time
+}
+
+func (e *serverEndpoint) handleDiscoControlMsg(from netip.AddrPort, senderIndex int, discoMsg disco.Message, uw udpWriter, serverDisco key.DiscoPublic) {
+	handshakeState := e.handshakeState[senderIndex]
+	if handshakeState == disco.BindUDPRelayHandshakeStateAnswerReceived {
+		// this sender is already bound
+		return
+	}
+	switch discoMsg := discoMsg.(type) {
+	case *disco.BindUDPRelayEndpoint:
+		switch handshakeState {
+		case disco.BindUDPRelayHandshakeStateInit:
+			// set sender addr
+			e.addrPorts[senderIndex] = from
+			fallthrough
+		case disco.BindUDPRelayHandshakeStateChallengeSent:
+			if from != e.addrPorts[senderIndex] {
+				// this is a later arriving bind from a different source, or
+				// a retransmit and the sender's source has changed, discard
+				return
+			}
+			m := new(disco.BindUDPRelayEndpointChallenge)
+			copy(m.Challenge[:], e.challenge[senderIndex][:])
+			reply := make([]byte, packet.GeneveFixedHeaderLength, 512)
+			gh := packet.GeneveHeader{Control: true, VNI: e.vni, Protocol: packet.GeneveProtocolDisco}
+			err := gh.Encode(reply)
+			if err != nil {
+				return
+			}
+			reply = append(reply, disco.Magic...)
+			reply = serverDisco.AppendTo(reply)
+			box := e.discoSharedSecrets[senderIndex].Seal(m.AppendMarshal(nil))
+			reply = append(reply, box...)
+			uw.WriteMsgUDPAddrPort(reply, nil, from)
+			// set new state
+			e.handshakeState[senderIndex] = disco.BindUDPRelayHandshakeStateChallengeSent
+			return
+		default:
+			// disco.BindUDPRelayEndpoint is unexpected in all other handshake states
+			return
+		}
+	case *disco.BindUDPRelayEndpointAnswer:
+		switch handshakeState {
+		case disco.BindUDPRelayHandshakeStateChallengeSent:
+			if from != e.addrPorts[senderIndex] {
+				// sender source has changed
+				return
+			}
+			if !bytes.Equal(discoMsg.Answer[:], e.challenge[senderIndex][:]) {
+				// bad answer
+				return
+			}
+			// sender is now bound
+			// TODO: Consider installing a fast path via netfilter or similar to
+			// relay (NAT) data packets for this serverEndpoint.
+			e.handshakeState[senderIndex] = disco.BindUDPRelayHandshakeStateAnswerReceived
+			// record last seen as bound time
+			e.lastSeen[senderIndex] = time.Now()
+			return
+		default:
+			// disco.BindUDPRelayEndpointAnswer is unexpected in all other handshake
+			// states, or we've already handled it
+			return
+		}
+	default:
+		// unexpected Disco message type
+		return
+	}
+}
+
+func (e *serverEndpoint) handleSealedDiscoControlMsg(from netip.AddrPort, b []byte, uw udpWriter, serverDisco key.DiscoPublic) {
```

Comment:
i may not have gotten far enough to see if this type gets reused in the clients, but if it's purely server side, it's giving me pause if the data oriented approach is the right shape for this, as we can generally infer the "sidedness" from the IO path. I like that the IO isn't coupled, but perhaps the state could be directly referenced from inferred knowledge, i.e. client and server states having pointers/field symbols rather than indices. I don't feel super strongly about it other than it'd be shame to have to do the extra comparisons here all the time when we can mostly infer the same fact from context.

Reply:
There is no client code besides what's in a unit test, so no potential re-use at the moment. Though I don't think this would be re-used, anyway.

There is natural coupling of client pairs (a relay forwards between a pair of clients using the same VNI), hence the indexing. A lookup using disco key pub key is unavoidable, though this could re-shaped to remove the arrays in trade for disco pub key lookup returning a pointer to a thing that is for just one of the clients, and they point at each other. I don't feel super strongly either way.